armeria-server/src/main/scala/org/http4s/server/armeria/ArmeriaHttp4sHandler.scala

```diff
@@ -0,0 +1,206 @@
+      serviceFn
+    else {
+      val timeoutResponse: F[Response[F]] = F.async[Response[F]] { cb =>
+        discardReturn {
+          ctx.log().whenComplete().thenRun(() => cb(Right(Response.timeout[F])))
+        }
+      }
+      // Use explicit an execution context to serve the service in Armeria's event loop
+      req => F.race(Async.shift(ec) *> serviceFn(req), timeoutResponse).map(_.merge)
+    }
+
+  private def onParseFailure(parseFailure: ParseFailure): F[HttpResponse] = {
+    val response = Response[F](Status.BadRequest).withEntity(parseFailure.sanitized)
+    F.pure(toHttpResponse(response))
+  }
+
+  /** Converts http4s' [[Response]] to Armeria's [[HttpResponse]]. */
+  private def toHttpResponse(response: Response[F]): HttpResponse = {
+    val headers = Stream(toResponseHeaders(response.headers, response.status.some))
+    val body: Stream[F, HttpData] = response.body.chunks.map { chunk =>
+      val byteChunk = chunk.toBytes
+      HttpData.copyOf(chunk.toBytes.values, byteChunk.offset, byteChunk.length)
+    }
+    val trailers = Stream
+      .eval(response.trailerHeaders)
+      .flatMap { trailers =>
+        if (trailers.isEmpty)
+          Stream.empty
+        else
+          Stream(toResponseHeaders(trailers, None))
+      }
+
+    HttpResponse.of((headers ++ body ++ trailers).toUnicastPublisher())
+  }
+
+  /** Converts Armeria's [[HttpRequest]] to http4s' [[Request]]. */
+  private def toRequest(ctx: ServiceRequestContext, req: HttpRequest): ParseResult[Request[F]] = {
+    val path = req.path()
+    for {
+      method <- Method.fromString(req.method().name())
+      uri <- Uri.requestTarget(path)
+    } yield Request(
+      method = method,
+      uri = uri,
+      httpVersion =
+        if (ctx.sessionProtocol().isMultiplex)
+          HttpVersion.`HTTP/2.0`
+        else if (req.headers().get(HttpHeaderNames.HOST) != null)
+          HttpVersion.`HTTP/1.1`
+        else
+          HttpVersion.`HTTP/1.0`,
+      headers = toHeaders(req),
+      body = toBody(req),
+      attributes = requestAttributes(ctx, uri)
+    )
+  }
+
+  /** Converts http4s' [[Headers]] to Armeria's [[ResponseHeaders]]. */
+  private def toResponseHeaders(headers: Headers, status: Option[Status]): ResponseHeaders = {
+    val builder = status
+      .map(s => ResponseHeaders.builder(s.code))
+      .getOrElse(ResponseHeaders.builder())
+
+    for (header <- headers.toList)
+      builder.add(header.name.toString, header.value)
+
+    builder.build()
+  }
+
+  /** Converts Armeria's [[HttpRequest]] to htt4s' [[Headers]]. */
+  private def toHeaders(req: HttpRequest): Headers =
+    Headers(
+      req
+        .headers()
+        .asScala
+        .map(entry => Header(entry.getKey.toString(), entry.getValue))
+        .toList
+    )
+
+  /** Converts a HTTP payload to [[EntityBody]]. */
+  private def toBody(req: HttpRequest): EntityBody[F] =
+    req
+      .toStream[F]
+      .filter(_.isInstanceOf[HttpData])
```

Comment:
can be simplified to 
`collect{x: HttpData => Chunk.bytes(x.array())}.flatMap(Stream.chunk)`

Reply:
That's what I want. Thanks!